### PR TITLE
feat: process the github_users.json file

### PR DESCRIPTION
**Pull request by hi-rustin**

### CODE REVIEW

**Comment by hi-rustin** [comment]
/hold

- [ ] tests

**Comment by hi-rustin** [comment]
/assgin @Mini256 

PTAL~ Thanks!

**Comment by hi-rustin** [comment]
/assign @Mini256 

PTAL~ Thanks!

File: tools/affiliations/importer/src/main.rs [hunk 1]
```diff
@@ -1,17 +1,369 @@
...
+            (
+                Some("".to_string()), Some("".to_string())
+            ),
+            (
+                Some("PerkinElmer < 2014-08-01, Independent < 2015-10-01, PwC < 2020-01-01, Simplebet".to_string()),
+                Some("PerkinElmer < 2014-08-01, Independent < 2015-10-01, PwC < 2020-01-01, Simplebet".to_string())
+            ),
+            (
+                Some("PingCAP".to_string()),
+                Some(format!("PingCAP < {}, Independent",current_date))
```
[the remark this answers is not in the document]
Comment (hi-rustin) [reply tools/affiliations/importer/src/main.rs:346]: It is to test if the date is correct.

File: tools/affiliations/importer/src/main.rs [hunk 2]
```diff
@@ -1,17 +1,369 @@
...
+                match affiliation
+                    .split(COMPANY_SEPARATOR)
+                    .collect::<Vec<&str>>()
+                    .last()
+                {
+                    None => {}
+                    Some(last) => {
+                        // Only if the last company is PingCAP.
+                        if *last == PINGCAP_AFFILIATION {
+                            let current_date = Utc::now().date().format("%Y-%m-%d").to_string();
```
Comment (hi-rustin) [inline tools/affiliations/importer/src/main.rs:141]: @Mini256 here

Comment (Mini256) [reply tools/affiliations/importer/src/main.rs:141]: But in fact, we didn't change the upstream JSON.

For example:
If we updated the affiliation information on May 15, The output is "PingCAP <2021-05-15, independent".

When updating the affiliation information while June 15, climbing JSON gets from the upstream JSON is still "PingCAP ", which will become "PingCAP <2021-06-15, independent" after remove_pingcap_affiliation processing.

This actor information may always be subordinate to PingCAP.

File: tools/affiliations/importer/src/main.rs [hunk 3]
```diff
@@ -1,17 +1,399 @@
...
+        let login = record.login.to_lowercase();
+        if logins.contains(&login) {
+            set_pingcap_affiliation(record);
+            processed_logins.insert(login);
+        } else {
+            remove_pingcap_affiliation(record);
+        }
+    }
+
+    // Processes the remaining logins.
```
Comment (Mini256) [inline tools/affiliations/importer/src/main.rs:90]:
```suggestion
    // Processes the remaining logins that existed in the logins but not in the github_logins file.
```

File: tools/affiliations/importer/src/main.rs [hunk 4]
```diff
@@ -1,17 +1,399 @@
...
+    // we need to process both, not just one and then consider the process complete.
+    let mut processed_logins: HashSet<String> = HashSet::new();
+    for record in &mut records {
+        // NOTICE: login is not case-sensitive.
+        let login = record.login.to_lowercase();
+        if logins.contains(&login) {
+            set_pingcap_affiliation(record);
+            processed_logins.insert(login);
+        } else {
+            remove_pingcap_affiliation(record);
```
Comment (Mini256) [inline tools/affiliations/importer/src/main.rs:86]: The record processed by `remove_pingcap_affiliation` function also needs to be inserted into the `processed_logins` set.

File: tools/affiliations/importer/src/main.rs [hunk 5]
```diff
@@ -1,17 +1,399 @@
...
+        .expect("Failed to list logins.");
+    let logins: HashSet<String> = logins
+        .iter()
+        .map(|l| l.trim())
+        // NOTICE: login is not case-sensitive.
+        .map(|l| l.to_lowercase())
+        .collect::<HashSet<_>>();
+    info!("Fetching the logins to succeed!");
+
+    // Because a login may appear twice in the original data and their affiliation may be different,
```
[the remark this answers is not in the document]
Comment (hi-rustin) [reply tools/affiliations/importer/src/main.rs:76]: Will our records appear multiple times?

File: tools/affiliations/importer/src/main.rs (hunk 3 above)
Comment (hi-rustin) [reply tools/affiliations/importer/src/main.rs:90]: Here are the rest of the logins in github_loings.

File: tools/affiliations/importer/src/main.rs (hunk 4 above)
Comment (hi-rustin) [reply tools/affiliations/importer/src/main.rs:86]: The main focus here is on github_logins, if it is not in github_logins then there is no need to record and add a new record.

File: tools/affiliations/importer/src/main.rs (hunk 5 above)
[the remark this answers is not in the document]
Comment (Mini256) [reply tools/affiliations/importer/src/main.rs:76]: Maybe it's possible.

https://github.com/cncf/devstatscode/blob/96c2b0551020d6a95f4cc876410f79481b921f3c/cmd/import_affs/import_affs.go#L736

And I found that "One email can also belong to multiple actors":

https://github.com/cncf/devstatscode/blob/96c2b0551020d6a95f4cc876410f79481b921f3c/cmd/import_affs/import_affs.go#L850

File: tools/affiliations/importer/src/main.rs (hunk 4 above)
Comment (Mini256) [reply tools/affiliations/importer/src/main.rs:86]: I understand.

**Review by Mini256 — APPROVED** [review]

**Comment by hi-rustin** [comment]
@Mini256 Thanks for your review!

**Pull request merged**